File: src/i1tbfssymmetric/Java/com/example/tbfssymmetric/BinaryTree.java
package com.example.tbfssymmetric;

import java.util.List;
import java.util.LinkedList;
import java.util.Queue;

public class BinaryTree<T> {
    public TreeNode<T> root;

    public BinaryTree(List<TreeNode<T>> listOfNodes) {
        this.root = createBinaryTree(listOfNodes);
    }

    private TreeNode<T> createBinaryTree(List<TreeNode<T>> listOfNodes) {
        if (listOfNodes.isEmpty()) return null;

        TreeNode<T> root = new TreeNode<>(listOfNodes.get(0).data);
        Queue<TreeNode<T>> queue = new LinkedList<>();
        queue.add(root);
        int i = 1;

        while (i < listOfNodes.size()) {
            TreeNode<T> curr = queue.poll();

            if (listOfNodes.get(i) != null) {
                curr.left = new TreeNode<>(listOfNodes.get(i).data);
                queue.add(curr.left);
            }
            i++;

            if (i < listOfNodes.size() && listOfNodes.get(i) != null) {
                curr.right = new TreeNode<>(listOfNodes.get(i).data);
                queue.add(curr.right);
            }
            i++;
        }

        return root;
    }
}

File: src/i1tbfssymmetric/CS/TBFSSymmetric.Algo/BinaryTree.cs
namespace TBFSSymmetric.Algo;

public class BinaryTree<T>
{
    public TreeNode<T>? Root { get; private set; }

    public BinaryTree(List<TreeNode<T>?> listOfNodes)
    {
        Root = CreateBinaryTree(listOfNodes);
    }

    private TreeNode<T>? CreateBinaryTree(List<TreeNode<T>?> listOfNodes)
    {
        if (listOfNodes.Count == 0) return null;

        var root = new TreeNode<T>(listOfNodes[0]!.Data);
        var queue = new Queue<TreeNode<T>>();
        queue.Enqueue(root);
        int i = 1;

        while (i < listOfNodes.Count)
        {
            var curr = queue.Dequeue();

            if (listOfNodes[i] != null)
            {
                curr.Left = new TreeNode<T>(listOfNodes[i]!.Data);
                queue.Enqueue(curr.Left);
            }
            i++;

            if (i < listOfNodes.Count && listOfNodes[i] != null)
            {
                curr.Right = new TreeNode<T>(listOfNodes[i]!.Data);
                queue.Enqueue(curr.Right);
            }
            i++;
        }

        return root;
    }
}
